src/Transports.AspNetCore/GraphQLHttpMiddlewareOptions.cs

```diff
@@ -0,0 +1,107 @@
+
+using System.Security.Claims;
+using System.Security.Principal;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+
+namespace GraphQL.Server.Transports.AspNetCore;
+
+/// <summary>
+/// Configuration options for <see cref="GraphQLHttpMiddleware"/>.
+/// </summary>
+public class GraphQLHttpMiddlewareOptions
+{
+    /// <summary>
+    /// Enables handling of GET requests.
+    /// </summary>
+    public bool HandleGet { get; set; } = true;
+
+    /// <summary>
+    /// Enables handling of POST requests, including form submissions, JSON-formatted requests and raw query requests.
+    /// </summary>
+    public bool HandlePost { get; set; } = true;
+
+    /********** WebSockets support ********
+    
+    /// <summary>
+    /// Enables handling of WebSockets requests.
+    /// <br/><br/>
+    /// Requires calling <see cref="WebSocketMiddlewareExtensions.UseWebSockets(IApplicationBuilder)"/>
+    /// to initialize the WebSocket pipeline within the ASP.NET Core framework.
+    /// </summary>
+    public bool HandleWebSockets { get; set; } = true;
+
+    ******************/
+
+    /// <summary>
+    /// Enables handling of batched GraphQL requests for POST requests when formatted as JSON.
+    /// </summary>
+    public bool EnableBatchedRequests { get; set; } = true;
+
+    /// <summary>
+    /// Enables parallel execution of batched GraphQL requests.
+    /// </summary>
+    public bool BatchedRequestsExecuteInParallel { get; set; } = true;
+
+    /// <summary>
+    /// When enabled, GraphQL requests with validation errors
+    /// have the HTTP status code set to 400 Bad Request.
+    /// GraphQL requests with execution errors are unaffected.
+    /// <br/><br/>
+    /// Does not apply to batched or WebSocket requests.
+    /// </summary>
+    public bool ValidationErrorsReturnBadRequest { get; set; } = true;
+
+    /// <summary>
+    /// Enables parsing the query string on POST requests.
+    /// If enabled, the query string properties override those in the body of the request.
+    /// </summary>
+    public bool ReadQueryStringOnPost { get; set; } = true;
+
+    /// <summary>
+    /// Enables reading variables from the query string.
+    /// Variables are interpreted as JSON and deserialized before being
+    /// provided to the <see cref="IDocumentExecuter"/>.
+    /// </summary>
+    public bool ReadVariablesFromQueryString { get; set; } = true;
+
+    /// <summary>
+    /// Enables reading extensions from the query string.
+    /// Extensions are interpreted as JSON and deserialized before being
+    /// provided to the <see cref="IDocumentExecuter"/>.
+    /// </summary>
+    public bool ReadExtensionsFromQueryString { get; set; } = true;
+
+    /// <summary>
+    /// If set, requires that <see cref="IIdentity.IsAuthenticated"/> return <see langword="true"/>
+    /// for the user within <see cref="HttpContext.User"/>
+    /// prior to executing the GraphQL request or accepting the WebSocket connection.
```

Comment:
executing in terms of _doing at least something even before validation_ ?

Reply:
"executing" in this context means calling `IDocumentExecuter`.  So these 3 validation rules occur prior to calling `IDocumentExecuter` and hence prior to parsing the query at all.  In this manner you can set an authorization rule for an endpoint separate from any configuration on the schema.  I would use this for my private applications that are exposed to the internet; if the JWT token is not validated, there is absolutely no processing of the query.  Without these options, a caller might be able to provide malformed queries to investigate the schema based on the types of responses received.  This is especially true as our error messages for missing fields actually suggests an alternative field - something like this is returned "could not find field 'dear' -- did you mean 'deer'?".  (They would of course receive authorization failures as well.)  But none of that is possible if the query does not execute to begin with.

Reply:
> So these 3 validation rules occur prior to calling IDocumentExecuter and hence prior to parsing the query at all

I see only 2 new rules - HttpGetValidationRule and HttpPostValidationRule and they are executed after query was parsed since they accept ValidationContext object obviosly. Or are you talking about these 3 configuration _properties_ used from within `AuthorizeAsync` helper method? I suppose the latter.

Regarding all that auth stuff here - it doesn't seem to you that such behavior is a task that does not relate to GraphQL middleware at all? It's some sort of a generic auth process that could (and should?) be baked into another middleware.

Reply:
> I suppose the latter.

Yes

> it doesn't seem to you that such behavior is a task that does not relate to GraphQL middleware at all

No it doesn't seem that way to me at all.  ASP.Net Core is closely tied to authentication for the entire pipeline.  There are abstractions to allow different authentication and authorization mechanisms -- in the case of Windows authentication, basic authentication, cookie authentication, OAuth authentication, JWT authentication, etc -- but the concept of authorization policies, roles, claims, and how they are set up to be used within ASP.Net Core is not configurable.  In the same manner, we should support the same basic concepts, allowing it to tie into the ASP.NET Core authorization and authentication frameworks.

Just as people do not need to apply ASP.NET Core authorization policies via attributes to ASP.NET Core controller actions, but can write their own attributes for custom behavior, nor does anyone need to use the authorization properties configured here in this middleware.  They can inherit from the class to add their own authorization properties, or implement separate middleware as necessary.

Futher, my plan (as I have implemented it in GraphQL.AspNetCore3) is to keep the authorization validation rule separate, requiring `builder.AddAuthorization()` to enable the authorization validation rule.  This allows for using the `GraphQL.Authorization` library or another authorization library for the metadata applied to the GraphQL schema, if so desired by the user.

It is also probably likely that if people are integrating GraphQL with an existing ASP.NET Core web server, they will likely use the same authentication interfaces that ASP.NET Core provides, which the Server project's validation rule will authorize against.  But it is common enough for custom authentication/authorization -- even my projects so far have custom middleware and custom authorization code.  But hopefully shortly my projects will be on the GraphQL.AspNetCore3 or maybe GraphQL.Server libraries.

Reply:
Note that I do not see any need to have the ASP.NET Core authorization rule as a separate project.  It is distinctly a part of ASP.NET Core.  It requires no additional dependencies.  People are not required to use it.  They are not in separate repos so we cannot release versions separately.  I just don't see a good reason to keep it separate.  In GraphQL.AspNetCore3, there is just a single nuget package, with a single dependency on ASP.Net Core which works for .NET Core 3.1 or newer frameworks.  We can discuss this later of course, and if it remains separate, so be it.

Reply:
OK to keep as is.